posts/discriminative-vs-generative/index.qmd

```diff
@@ -0,0 +1,598 @@
+date: "2022/04/03"
+image: "graph.png"
+categories: [factor graphs, PGMs, ML, theory]
+echo: false
+jupyter:
+  jupytext:
+    text_representation:
+      extension: .qmd
+      format_name: quarto
+      format_version: '1.0'
+      jupytext_version: 1.13.7
+  kernelspec:
+    display_name: Python 3.6.15 ('fgm')
+    language: python
+    name: python3
+---
+
+```{python}
+#| eval: false
+!pip install -r requirements.txt
+```
+
+```{python}
+from graphviz import Source
+import matplotlib.pyplot as plt
+```
+
+```{python}
+plt.rcParams['text.usetex'] = True
+```
+
+Looking around, there seem to be multiple definitions of what discriminative and generative models actually are, ranging from too narrow to outright confusing, as well as lots of claims about advantages and disadvantages of one over the other, but often with little detail or justification. 
+However, if one ignores the miriad references on this topic and just reads through @bishop2006pattern [Chapter 4] and @minka2005discriminative, one gets a fairly good picture of what's what.
+
+We will work through the definitions and differences by considering _probabilistic (graphical) models_, specifically using the _factor graph_ representation[^1].
+Throughout, we will focus on _classification_, where we are given a labelled data set $D = \lbrace \left( x^i, y^i \right) \rbrace_N$ with inputs $x^i = \left( x^i_1, \cdots, x^i_n \right)$ and labels $y^i \in \lbrace 0, 1 \rbrace$, and we want $P \left( y = 1 \vert x = \hat x \right)$ for some (new) $\hat x$[^2].
+We should keep in mind that the concepts and the generative-discriminative dichotomy are more general, i.e. they are not restricted to the problem of classification.
+
+[^1]: Factor graphs are a more detailed representation of a probabilistic graphical model that specifies the factorization of the joint probability being modelled.
+See e.g. @loeliger2004introduction.
+
+[^2]: Note that actually determining which class to assign to $\hat x$ requires a final decision. 
+We will ignore this and _discriminant functions_, which combine the inference and the decision steps.
+See [@bishop2006pattern, Section 1.5.4] for a discussion.
+
+## Generative models can be used to generate synthetic data
+
+Many references jump straight in and define deterministic and generative models, in the context of classification, via the probabilities that they consider. 
+A _discriminative model_ is defined as one that provides $P \left( y \vert x \right)$ directly.
+The name comes from the fact that we can view the distribution (together with the decision to choose the most probable value) as directly discriminating the value of the target $y$ for any given instance $x$.
+
+```{python}
+#| fig-cap: Discriminative model, as used for prediction.
+
+model_discriminative_str = """
+strict digraph {
+    graph [rankdir=LR];
+    # factors
+    node [shape=square];
+    p_y_x [label="P(y|x)"];
```

Comment:
in order to be consistent with later sections, this should be X and Y, given we cant do \hat x and \hat y. 

i think this is unnecessary, and ugly, and say we keep as is, despite the minor inconsistency. it does annoy me a bit though.

alternatively we don't say it's being used for prediction, and ungrey the xs.